File: src/main/java/rs/helpkit/pref/RSPreferences.java
package rs.helpkit.pref;

import java.awt.*;
import java.io.*;
import java.util.Properties;

/**
 * @author Tyler Sedlar
 * @since 3/20/2018
 */
public class RSPreferences {

    private static final Properties PROPS = new Properties();

    private interface Keys {
        String DEFAULT_WORLD = "default_world";
        String DEFAULT_WIDTH = "default_width";
        String DEFAULT_HEIGHT = "default_height";
        String DEFAULT_X = "default_x";
        String DEFAULT_Y = "default_y";
    }

    static {
        File propertyFile = new File(HKConfig.HOME, "prefs.txt");
        if (propertyFile.exists()) {
            try (InputStream in = new FileInputStream(propertyFile)) {
                PROPS.load(in);
            } catch (IOException e) {
                System.err.println("Failed to load prefs.txt");
            }
        } else {
        }
        setIfInvalid(Keys.DEFAULT_WORLD, "70");
        setIfInvalid(Keys.DEFAULT_WIDTH, "765");
        setIfInvalid(Keys.DEFAULT_HEIGHT, "503");
        setIfInvalid(Keys.DEFAULT_X, "-1");
        setIfInvalid(Keys.DEFAULT_Y, "-1");
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            try (OutputStream out = new FileOutputStream(propertyFile)) {
                PROPS.store(out, "RSHelpKit preferences");
            } catch (IOException e) {
                e.printStackTrace();
            }
        }));
    }

    private static void setIfInvalid(String key, String value) {
        if (PROPS.getProperty(key) == null) {
            PROPS.setProperty(key, value);
        }
    }

    public static Dimension getDefaultSize() {
        int width = Integer.parseInt(PROPS.get(Keys.DEFAULT_WIDTH).toString());
        int height = Integer.parseInt(PROPS.get(Keys.DEFAULT_HEIGHT).toString());
        return new Dimension(width, height);
    }

    public static void setDefaultSize(int width, int height) {
        PROPS.setProperty(Keys.DEFAULT_WIDTH, Integer.toString(width));
        PROPS.setProperty(Keys.DEFAULT_HEIGHT, Integer.toString(height));
    }

    public static Point getDefaultLocation() {
        int x = Integer.parseInt(PROPS.get(Keys.DEFAULT_X).toString());
        int y = Integer.parseInt(PROPS.get(Keys.DEFAULT_Y).toString());
        return new Point(x, y);
    }

    public static void setDefaultLocation(int x, int y) {
        PROPS.setProperty(Keys.DEFAULT_X, Integer.toString(x));
        PROPS.setProperty(Keys.DEFAULT_Y, Integer.toString(y));
    }

    public static int getDefaultWorld() {
        return Integer.parseInt(PROPS.get(Keys.DEFAULT_WORLD).toString());
    }

    public static void setDefaultWorld(int world) {
        PROPS.setProperty(Keys.DEFAULT_WORLD, Integer.toString(world));
    }
}

File: src/main/java/rs/helpkit/pref/RSPreferences.kt
package rs.helpkit.pref

import rs.helpkit.api.game.access.Camera
import rs.helpkit.api.game.access.Skills
import java.awt.Dimension
import java.awt.Point
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.util.*

/**
 * @author Tyler Sedlar
 * @since 3/20/2018
 */
object RSPreferences {

    private val PROPS = Properties()
    private val XP_PROPS = Properties()

    val defaultSize: Dimension
        get() {
            val width = Integer.parseInt(PROPS[Keys.DEFAULT_WIDTH].toString())
            val height = Integer.parseInt(PROPS[Keys.DEFAULT_HEIGHT].toString())
            return Dimension(width, height)
        }

    val defaultLocation: Point
        get() {
            val x = Integer.parseInt(PROPS[Keys.DEFAULT_X].toString())
            val y = Integer.parseInt(PROPS[Keys.DEFAULT_Y].toString())
            return Point(x, y)
        }

    var defaultWorld: Int
        get() = Integer.parseInt(PROPS[Keys.DEFAULT_WORLD].toString())
        set(world) {
            PROPS.setProperty(Keys.DEFAULT_WORLD, Integer.toString(world))
        }

    var defaultZoomLevel: Int
        get() = Integer.parseInt(PROPS[Keys.DEFAULT_ZOOM_LEVEL].toString())
        set(zoom) {
            PROPS.setProperty(Keys.DEFAULT_ZOOM_LEVEL, Integer.toString(zoom))
        }

    private interface Keys {
        companion object {
            const val DEFAULT_WORLD = "default_world"
            const val DEFAULT_WIDTH = "default_width"
            const val DEFAULT_HEIGHT = "default_height"
            const val DEFAULT_X = "default_x"
            const val DEFAULT_Y = "default_y"
            const val DEFAULT_ZOOM_LEVEL = "default_zoom_level"
        }
    }

    private fun loadProps(file: File, props: Properties) {
        if (file.exists()) {
            try {
                FileInputStream(file).use { props.load(it) }
            } catch (e: IOException) {
                System.err.println("Failed to load prefs.txt")
            }
        }
    }

    init {
        val mainProps = File(HKConfig.HOME, "prefs.txt")
        val xpProps = File(HKConfig.HOME, "xp_prefs.txt")

        loadProps(mainProps, PROPS)
        loadProps(xpProps, XP_PROPS)

        // set default PROPS
        setIfInvalid(PROPS, Keys.DEFAULT_WORLD, "70")
        setIfInvalid(PROPS, Keys.DEFAULT_WIDTH, "765")
        setIfInvalid(PROPS, Keys.DEFAULT_HEIGHT, "503")
        setIfInvalid(PROPS, Keys.DEFAULT_X, "-1")
        setIfInvalid(PROPS, Keys.DEFAULT_Y, "-1")
        setIfInvalid(PROPS, Keys.DEFAULT_ZOOM_LEVEL, Camera.ZOOM_OUT_MAX.toString())

        // set default XP_PROPS
        Skills.values().forEach {
            val name = it.skillName().toLowerCase()
            setIfInvalid(XP_PROPS, "${name}_enabled", "false")
            setIfInvalid(XP_PROPS, "${name}_frame_xpos", "-1")
            setIfInvalid(XP_PROPS, "${name}_frame_ypos", "-1")
            setIfInvalid(XP_PROPS, "${name}_frame_popped", "false")
        }

        Runtime.getRuntime().addShutdownHook(Thread {
            try {
                FileOutputStream(mainProps).use { out -> PROPS.store(out, "RSHelpKit preferences") }
                FileOutputStream(xpProps).use { out -> XP_PROPS.store(out, "XP plugin preferences") }
            } catch (e: IOException) {
                e.printStackTrace()
            }
        })
    }

    private fun setIfInvalid(props: Properties, key: String, value: String) {
        if (props.getProperty(key) == null) {
            props.setProperty(key, value)
        }
    }

    fun setDefaultSize(width: Int, height: Int) {
        PROPS.setProperty(Keys.DEFAULT_WIDTH, Integer.toString(width))
        PROPS.setProperty(Keys.DEFAULT_HEIGHT, Integer.toString(height))
    }

    fun setDefaultLocation(x: Int, y: Int) {
        PROPS.setProperty(Keys.DEFAULT_X, Integer.toString(x))
        PROPS.setProperty(Keys.DEFAULT_Y, Integer.toString(y))
    }

    fun isSkillEnabled(skill: Skills): Boolean {
        val name = skill.skillName().toLowerCase()
        return XP_PROPS.getProperty("${name}_enabled")!!.toBoolean()
    }

    fun setSkillEnabled(skill: Skills, enabled: Boolean) {
        val name = skill.skillName().toLowerCase()
        XP_PROPS.setProperty("${name}_enabled", enabled.toString())
    }

    fun skillLocationFor(skill: Skills): Point {
        val name = skill.skillName().toLowerCase()
        val x = XP_PROPS.getProperty("${name}_frame_xpos").toInt()
        val y = XP_PROPS.getProperty("${name}_frame_ypos").toInt()
        return Point(x, y)
    }

    fun setSkillLocation(skill: Skills, x: Int, y: Int) {
        val name = skill.skillName().toLowerCase()
        XP_PROPS.setProperty("${name}_frame_xpos", x.toString())
        XP_PROPS.setProperty("${name}_frame_ypos", y.toString())
    }

    fun isSkillPopout(skill: Skills): Boolean {
        val name = skill.skillName().toLowerCase()
        return XP_PROPS.getProperty("${name}_frame_popped")!!.toBoolean()
    }

    fun setSkillPopped(skill: Skills, popped: Boolean) {
        val name = skill.skillName().toLowerCase()
        XP_PROPS.setProperty("${name}_frame_popped", popped.toString())
    }
}
